### PR TITLE
CI endpoint

### PR DESCRIPTION
**Pull request by marcinwyszynski**
@thecodebeat/developers PTAL

### CODE REVIEW

File: internal/project_event.go [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+	"fmt"
+	"math"
+)
+
+type ProjectEvents []ProjectEvent
+
+func (pe ProjectEvents) delta() float64 {
+	for _, evt := range pe {
+		if evt.isProjectEvent() {
+			return evt.delta()
```
Comment (kruszczynski) [inline internal/project_event.go]: What if more than one `pe` has a `delta` shouldn't we collect these deltas?

File: internal/router.go [hunk 2]
```diff
@@ -40,3 +40,22 @@ func (r *router) projectTarballUpload(projectUUID string, data io.Reader) (*http
...
+	)
+	request, err := http.NewRequest("GET", urlStr, nil)
+	if err != nil {
+		return nil, err
+	}
+	request.Header.Add(tokenHeader, r.apiToken)
+	return request, nil
+}
+
+func (r *router) projectCommitUpload(request *commitRequest, data io.Reader) (*http.Request, error) {
```
Comment (kruszczynski) [inline internal/router.go]: I'm presuming that's why it's `WIP` ;)

File: internal/project_event.go (hunk 1 above)
Comment (marcinwyszynski) [inline internal/project_event.go]: We should only get max one such `ProjectEvent` (`gpa_degraded` or `gpa_improved`).

File: internal/router.go (hunk 2 above)
Comment (marcinwyszynski) [inline internal/router.go]: Correct.

File: internal/router.go [hunk 3]
```diff
@@ -56,6 +61,20 @@ func (r *router) projectCommitShow(projectUUID string, commitUUID string) (*http
...
+		"%s/cli/projects/%s/commits",
+		strings.TrimRight(r.endpoint, "/"),
+		req.projectUUID,
+	)
+	request, err := http.NewRequest("POST", urlStr, data)
+	if err != nil {
+		return nil, err
+	}
+	request.Header.Add(tokenHeader, r.apiToken)
+	request.Header.Add("Content-Type", "application/octet-stream")
```
Comment (kruszczynski) [inline internal/router.go]: What's special about `Content-Type` that it's not a header?

Comment (marcinwyszynski) [inline internal/router.go]: It's not application-defined.

File: internal/project_event.go (hunk 1 above)
Comment (grodowski) [inline internal/project_event.go]: It looks like `isProjectEvent` is an uncommunicative method name (every `pe` of `ProjectEvents` is a Project Event). How about `isProjectGpaEvent`?

Comment (marcinwyszynski) [inline internal/project_event.go]: Done.

**Comment by kruszczynski** [comment]
✅

File: internal/project_upload_commit.go [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+	}
+	if response.StatusCode != http.StatusOK && response.StatusCode != http.StatusAccepted {
+		fmt.Printf("server responded with status code %d\n", response.StatusCode)
+		return fmt.Errorf("server responded with status code %d", response.StatusCode)
+	}
+	if response.StatusCode == http.StatusAccepted {
+		time.Sleep(30 * time.Second)
+		return pollForCommitID(ctx, commitID)
+	}
+	defer response.Body.Close()
```
Comment (kruszczynski) [inline internal/project_upload_commit.go]: Don't we want `response.Body` to close even in case of `response.StatusCode == http.StatusAccepted`?

Comment (kruszczynski) [inline internal/project_upload_commit.go]: Or because we're not reading from it it does not apply?

Comment (marcinwyszynski) [inline internal/project_upload_commit.go]: Exactly. If we're not reading it we can ignore it.

**Comment by kruszczynski** [comment]
👍

**Pull request merged**